plugins/spi/SPIOutput.cpp

```diff
@@ -653,6 +661,91 @@ uint8_t SPIOutput::CalculateAPA102LatchBytes(uint16_t pixel_count) {
   return latch_bytes;
 }
 
+void SPIOutput::IndividualWS2812bControl(const DmxBuffer &buffer) {
+  const unsigned int first_slot = m_start_address - 1;  // 0 offset
+  if (buffer.Size() - first_slot < WS2812b_SLOTS_PER_PIXEL) {
+    OLA_INFO << "Insufficient DMX data, required " << WS2812b_SLOTS_PER_PIXEL
+             << ", got " << buffer.Size() - first_slot;
+    return;
+  }
+
+  // We always check out the entire string length, even if we only have data
+  // for part of it
+  const unsigned int output_length = m_pixel_count * WS2812b_SLOTS_PER_PIXEL;
+  uint8_t *output = m_backend->Checkout(m_output_number, output_length);
+  if (!output) {
+    OLA_INFO << "Unable to create output buffer of required length: " << output_length;
+    return;
+  }
+
+  const unsigned int length = std::min(m_pixel_count * WS2812b_SLOTS_PER_PIXEL,
+                                       buffer.Size() - first_slot);
+
+  for (unsigned int i = 0; i < length / WS2812b_SLOTS_PER_PIXEL; i++) {
+    // Convert RGB to GRB
+    unsigned int offset = first_slot + i * WS2812b_SLOTS_PER_PIXEL;
+    uint8_t r = buffer.Get(offset);
+    uint8_t g = buffer.Get(offset + 1);
+    uint8_t b = buffer.Get(offset + 2);
+    output[i * WS2812b_SPI_BYTES_PER_PIXEL] = 0x24 | ((g & 0x1) << 1) | ((g & 0x2) << 3) | ((g & 0x4) << 5);
```

Comment:
Given this code is used twice, it would be good to have a function that does it, passing in the DMX data and a reference to the array we want to write it into.

Reply:
The conversion code is the same thing 6 times over, wouldn't it be better to convert that part to a function. I struggle with passing arrays around in C++ as I don't really know C++, could you give me a hint on doing this?